File: android/src/main/java/com/mostafataghipour/reactnativeuploadmanager/UploadQueue.kt
package com.mostafataghipour.reactnativeuploadmanager

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject

interface UploadQueue {
    fun clear()
    fun push(json: JSONObject)
    fun pop(removeIt: Boolean): JSONObject?
    fun isEmpty(): Boolean
    fun remove(id: String)
}


class UploadQueueImp private constructor(context: Context) : UploadQueue {

    companion object {
        private const val PREF_KEY = "RN-UPLOAD_KEY"
        private const val QUEUE_KEY = "RN-UPLOAD-QUEUE_KEY"


        @Volatile private var INSTANCE: UploadQueueImp? = null

        fun getInstance(context: Context): UploadQueueImp =
                INSTANCE ?: synchronized(this) {
                    INSTANCE ?: buildInstance(context).also { INSTANCE = it }
                }

        private fun buildInstance(context: Context) =
                UploadQueueImp(context)
    }



    private val pref: SharedPreferences by lazy {
        return@lazy context.getSharedPreferences(PREF_KEY, Context.MODE_PRIVATE)
    }


    private lateinit var queue: JSONArray

    init {
        load()
    }

    private fun load() {
        val str = pref.getString(QUEUE_KEY, null)
        val jsonArray: JSONArray = if (str.isNullOrEmpty())
            JSONArray()
        else
            JSONArray(str)

        this.queue = jsonArray
    }

    override fun push(json: JSONObject) {
        this.queue.put(json)
        sync()
    }


    override fun pop(removeIt: Boolean): JSONObject? = try {
        val json = this.queue.getJSONObject(0)
        if (removeIt) {
            this.queue.remove(0)
            sync()
        }
        json
    } catch (e: Exception) {
        null
    }


    override fun remove(id: String) {
        try {
            for (i in 0 until this.queue.length()) {
                val item = this.queue.getJSONObject(i)
                if (item.getString("customUploadId") == id) {
                    this.queue.remove(i)
                    sync()
                    break
                }
            }
        } catch (e: Exception) {

        }
    }


    private fun sync() {
        val editor = pref.edit()
        val str = this.queue.toString()
        editor.putString(QUEUE_KEY, str)
        editor.apply()
    }

    override fun clear() {
        this.queue = JSONArray()
        sync()
    }

    override fun isEmpty(): Boolean {
        return this.queue.length() == 0
    }
}

File: ios/UploadQueue.swift
//
//  UploadQueue.swift
//  RNUploadManager
//
//  Created by Mostafa Taghipour on 2/17/21.
//  Copyright © 2021 Mostafa Taghipour. All rights reserved.
//

import Foundation

typealias UploadItem = [AnyHashable:Any]

protocol UploadQueue {
    func push(item: UploadItem)
    func pop(removeIt: Bool) -> UploadItem?
    func remove(id: String)
    func clear()
    func isEmpty() -> Bool
}


class UploadQueueImp: UploadQueue {
    
    static let shared = UploadQueueImp()
    
    private static let QUEUE_KEY = "UPLOAD_MANAGER_QUEUE_KEY"
    
    private lazy var userDefaults : UserDefaults = {
        return UserDefaults.standard
    }()
    
    var queue : [UploadItem] = [[:]]
    
    
    private init(){
        load()
    }
    
    
    private func load(){
        self.queue = userDefaults.object(forKey: UploadQueueImp.QUEUE_KEY) as? [UploadItem] ?? [[:]]
    }
    
    func push(item: UploadItem) {
        self.queue.append(item)
        sync()
    }
    
    
    func pop(removeIt: Bool) -> UploadItem? {
        let first =  self.queue.first
        
        if removeIt , first != nil {
            self.queue.remove(at: 0)
            sync()
        }
        
        return first
    }
    
    
    func remove(id: String) {
        let index = self.queue.firstIndex{$0["customUploadId"] as? String == id}
        
        if let index = index {
            self.queue.remove(at: index)
            sync()
        }
    }
    
    private func sync() {
        userDefaults.set(self.queue, forKey: UploadQueueImp.QUEUE_KEY)
    }
    
    func clear() {
        self.queue = []
        sync()
    }
    
    func isEmpty() -> Bool {
        return self.queue.isEmpty
    }
}
